src/algorithms/linear_stretching.jl

```diff
@@ -39,44 +43,149 @@ channel are stretched to the specified range. The modified Y channel is then
+        if !isnothing(minval)
+            dst_minval = minval
+            Base.depwarn("deprecated: use `dst_minval` for keyword `minval`", :LinearStretching)
+        end
+        if !isnothing(maxval)
+            dst_maxval = maxval
+            Base.depwarn("deprecated: use `dst_maxval` for keyword `maxval`", :LinearStretching)
+        end
+
+        dst_minval <= dst_maxval || throw(ArgumentError("dst_minval $dst_minval should be less than dst_maxval $dst_maxval"))
+        if !(isnothing(src_minval) || isnothing(src_maxval))
+            src_minval <= src_maxval || throw(ArgumentError("src_minval $src_minval should be less than src_maxval $src_maxval"))
+        end
+        T = promote_type(T1, T2, T3, T4, T5, T6)
+        new{T}(convert(T, src_minval), convert(T, src_maxval),
+               convert(T, dst_minval), convert(T, dst_maxval),
+               convert(T, dst_minval), convert(T, dst_maxval))
+    end
+end
+function LinearStretching(rangemap::Pair{Tuple{T1, T2}, Tuple{T3, T4}}) where {T1, T2, T3, T4}
+    LinearStretching(rangemap.first..., rangemap.second...)
+end
+function LinearStretching(rangemap::Pair{Nothing, Tuple{T3, T4}}) where {T3, T4}
+    LinearStretching(nothing, nothing, rangemap.second...)
+end
+function LinearStretching(rangemap::Pair{Tuple{T1, T2}, Nothing}) where {T1, T2}
+    LinearStretching(src_minval=rangemap.first[1], src_maxval=rangemap.first[2])
 end
 
 function (f::LinearStretching)(out::GenericGrayImage, img::GenericGrayImage)
-    src_minval = minfinite(img)
-    src_maxval = maxfinite(img)
     T = eltype(out)
-    out .= img
-    map!(out,out) do val
+    FT = eltype(floattype(T))
+    img_min, img_max = minfinite(img), maxfinite(img)
+    # explicit annotation is needed because the ?: line mixes three value types:
+    # Nothing, T, and typeof(f.src_minval)
+    src_minval::FT = isnothing(f.src_minval) ? img_min : f.src_minval
+    src_maxval::FT = isnothing(f.src_maxval) ? img_max : f.src_maxval
+    dst_minval::FT = f.dst_minval
+    dst_maxval::FT = f.dst_maxval
+
+    # the kernel operation `r * x - o` is equivalent to `(x-A) * ((b-a)/(B-A)) + a`
+    # precalculate these and make inner loop contains only multiplication and addition
+    # to get better performance
+    r = (dst_maxval - dst_minval) / (src_maxval - src_minval)
+    o = (src_minval*dst_maxval - src_maxval*dst_minval) / (src_maxval - src_minval)
+
+    if 1 ≈ r && 0 ≈ o
+        # when image intensity is already adjusted, there's no need to do it again
+        # it's a trivial but common case in practice
+        out === img || (out .= img)
+        return out
+    end
+
+    # In most cases, we don't need to clamp the output
+    # this is only used when user specifies custom `(src_minval, src_maxval)`
+    out_minval = r * img_min - o
+    out_maxval = r * img_max - o
+    do_clamp = (out_minval < dst_minval) || (out_maxval > dst_maxval)
```

Comment:
this PR is great as is, and is all i (at least currently) need.  however, making the clamp optional would provide for maximum flexibility.  would this be possible with a keyword argument, or otherwise?

thanks for all the hard work @johnnychen94

Reply:
In the first two commits, it's hard to handle this correctly wrt `FixedPoint`, and then I totally forget about this. Sorry about that.

I've added a `no_clamp` option in a254af4